File: backend/app/models/task.py
"""Domain models for academic tasks."""
from datetime import datetime
from dataclasses import dataclass, field
from typing import Optional


@dataclass(frozen=True)
class Task:
    """Represents an academic task to be scheduled."""
    id: str
    task_name: str
    credit_weight: int       # 1 to 6
    difficulty_score: int    # 1 to 10
    deadline: datetime
    study_duration_hours: float

    def __post_init__(self):
        if not self.id:
            raise ValueError("Task id cannot be empty.")
        if not self.task_name:
            raise ValueError("Task name cannot be empty.")
        if not (1 <= self.credit_weight <= 6):
            raise ValueError(f"Credit weight must be between 1 and 6, got {self.credit_weight}.")
        if not (1 <= self.difficulty_score <= 10):
            raise ValueError(f"Difficulty score must be between 1 and 10, got {self.difficulty_score}.")
        if self.study_duration_hours <= 0:
            raise ValueError(f"Study duration must be positive, got {self.study_duration_hours}.")

    @property
    def required_blocks(self) -> int:
        """Returns the required number of 15-minute time blocks."""
        return int(round(self.study_duration_hours * 4))

File: frontend/lib/models/task.dart
/// Academic task domain model.
///
/// Mirrors the backend TaskSchema (credit_weight: 1–6, difficulty_score: 1–10).
class Task {
  final String id;
  final String taskName;
  final int creditWeight;
  final int difficultyScore;
  final DateTime deadline;
  final double studyDurationHours;

  const Task({
    required this.id,
    required this.taskName,
    required this.creditWeight,
    required this.difficultyScore,
    required this.deadline,
    required this.studyDurationHours,
  });

  /// Number of 15-minute blocks required.
  int get requiredBlocks => (studyDurationHours * 4).ceil();

  Task copyWith({
    String? id,
    String? taskName,
    int? creditWeight,
    int? difficultyScore,
    DateTime? deadline,
    double? studyDurationHours,
  }) {
    return Task(
      id: id ?? this.id,
      taskName: taskName ?? this.taskName,
      creditWeight: creditWeight ?? this.creditWeight,
      difficultyScore: difficultyScore ?? this.difficultyScore,
      deadline: deadline ?? this.deadline,
      studyDurationHours: studyDurationHours ?? this.studyDurationHours,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'task_name': taskName,
      'credit_weight': creditWeight,
      'difficulty_score': difficultyScore,
      'deadline': deadline.toIso8601String(),
      'study_duration_hours': studyDurationHours,
    };
  }

  factory Task.fromJson(Map<String, dynamic> json) {
    return Task(
      id: json['id'] as String,
      taskName: json['task_name'] as String,
      creditWeight: json['credit_weight'] as int,
      difficultyScore: json['difficulty_score'] as int,
      deadline: DateTime.parse(json['deadline'] as String),
      studyDurationHours: (json['study_duration_hours'] as num).toDouble(),
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Task && runtimeType == other.runtimeType && id == other.id;

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() => 'Task(id: $id, taskName: $taskName)';
}
